src/mokernels/moinput.jl

```diff
@@ -18,40 +47,83 @@ julia> MOInput(x, 2)
  (3, 2)
 ```
 
-As shown above, an `MOInput` represents a vector of tuples.
+Accommodates modelling multi-dimensional output data where all outputs are always observed.
+
+As shown above, an `MOInputIsotopicByOutputs` represents a vector of tuples.
 The first `length(x)` elements represent the inputs for the first output, the second
 `length(x)` elements represent the inputs for the second output, etc.
-
-See [Inputs for Multiple Outputs](@ref) in the docs for more info.
 """
-struct MOInput{T<:AbstractVector} <: AbstractVector{Tuple{Any,Int}}
+struct MOInputIsotopicByOutputs{S,T<:AbstractVector{S}} <: AbstractVector{Tuple{S,Int}}
     x::T
-    out_dim::Integer
+    out_dim::Int
 end
 
-Base.length(inp::MOInput) = inp.out_dim * length(inp.x)
-
-Base.size(inp::MOInput, d) = d::Integer == 1 ? inp.out_dim * size(inp.x, 1) : 1
-Base.size(inp::MOInput) = (inp.out_dim * size(inp.x, 1),)
-
-Base.lastindex(inp::MOInput) = length(inp)
-Base.firstindex(inp::MOInput) = 1
-
-function Base.getindex(inp::MOInput, ind::Integer)
-    if ind > 0
-        out_dim = ind ÷ length(inp.x) + 1
-        ind = ind % length(inp.x)
-        if ind == 0
-            ind = length(inp.x)
-            out_dim -= 1
-        end
-        return (inp.x[ind], out_dim::Int)
-    else
-        throw(BoundsError(string("Trying to access at ", ind)))
-    end
+const IsotopicMOInputs = Union{MOInputIsotopicByFeatures,MOInputIsotopicByOutputs}
+
+function Base.getindex(inp::MOInputIsotopicByOutputs, ind::Integer)
+    @boundscheck checkbounds(inp, ind)
+    output_index, feature_index = fldmod1(ind, length(inp.x))
+    feature = @inbounds inp.x[feature_index]
+    return feature, output_index
 end
 
-Base.iterate(inp::MOInput) = (inp[1], 1)
-function Base.iterate(inp::MOInput, state)
-    return (state < length(inp)) ? (inp[state + 1], state + 1) : nothing
+function Base.getindex(inp::MOInputIsotopicByFeatures, ind::Integer)
+    @boundscheck checkbounds(inp, ind)
+    feature_index, output_index = fldmod1(ind, inp.out_dim)
+    feature = @inbounds inp.x[feature_index]
+    return feature, output_index
 end
+
+Base.size(inp::IsotopicMOInputs) = (inp.out_dim * length(inp.x),)
+
+"""
+    MOInput(x::AbstractVector, out_dim::Integer)
```

Comment:
```suggestion
    MOInput(x::AbstractVector, out_dim::Integer)

```

Also, would it be good to mention here that this is just for backwards compatibility? (Or otherwise clarify relation to MOInputIsotopicByOutputs)

Reply:
Doesn't using those argument names make this breaking?

Reply:
Definitely can add that. Just since it isn't exported it won't be in the docs so they will have to look at the source code docstring to see what the `MOInputIsotopicByOutputs` is if its mentioned

Reply:
Should we hold off on mentioning the backwards compatibility until we explicitly make the other types public by exporting them (or helper functions that build them)?

Reply:
Probably not a bad idea to add a note that says "as of version ... this feature is deprecated, and will be removed in the next breaking release"

Reply:
But isn't it not deprecated at the moment? If we deprecate it now, we have no public multi output input types. 

Also is the plan to have it completely removed? I thought the idea was to leave it as an alias to minimize breaking changes but I may have misunderstood that it was just a short term plan

Reply:
Note added, resolving